File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.py
def solution(genres, plays):
    play = []
    for number in range(len(plays)):
        play.append((number, plays[number]))

    all_gen = {}
    for i in range(len(genres)):

        if genres[i] in all_gen:
            all_gen[genres[i]].append(play[i])

        else:
            all_gen[genres[i]] = [play[i]]


    # 장르별 재생합
    genre_play_counts = {}
    for genre in all_gen:

        play_list = all_gen[genre]
        play_sum = 0
        for _,counts in play_list:
            play_sum += counts

        # play_sum = sum(counts for _, counts in play_list)

        genre_play_counts[genre] = play_sum



    top_genre = ""
    max_count = 0
    for i in genre_play_counts:
        if max_count < genre_play_counts[i]:
            max_count = genre_play_counts[i]
            top_genre = i


    answer = []
    top_sorted_list = sorted(all_gen[top_genre], key=lambda x:x[1], reverse=True)

    for i in range(min(2,len(top_sorted_list))):
        answer.append(top_sorted_list[i][0])

    sorted_genre_play_counts = sorted(genre_play_counts.items(), key=lambda x:x[1], reverse=True)
    for genre in sorted_genre_play_counts:
        if genre[0] == top_genre:
            pass
        else:
            sorted_list = sorted(all_gen[genre[0]], key=lambda x:x[1], reverse=True)
            for i in range(min(2, len(sorted_list))):
                answer.append(sorted_list[i][0])
                
    return answer

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.cpp
#include <string>
#include <vector>
#include <unordered_map>
#include <algorithm>

using namespace std;

bool compareGenre(const pair<string, int> &a, const pair<string, int> &b)
{
    return a.second > b.second;
}

bool comparePlay(const pair<int, int> &a, const pair<int, int> &b)
{
    // 같으면 번호가 더 빠른 것을 출력
    if(a.second == b.second)
    {
        return a.first < b.first;
    }
    return a.second > b.second;
}

vector<int> solution(vector<string> genres, vector<int> plays) {
    vector<int> answer;
    
    // 장르별 총 재생횟수
    unordered_map<string, int> genre_cnt;
    // 장르? 번호? 재생횟수?
    // pop : {1, 300}, {2, 400}
    unordered_map<string, vector<pair<int, int>>> genre_plays;
    
    // 장르별 수를 세기
    for(int i=0; i<genres.size(); i++)
    {
        genre_plays[genres[i]].push_back({i, plays[i]});
        genre_cnt[genres[i]] += plays[i];
    }
    
    // 정렬(100개정도는 변환 가능)
    vector<pair<string ,int>> sorted_genres(genre_cnt.begin(), genre_cnt.end());
    sort(sorted_genres.begin(), sorted_genres.end(), compareGenre);
    
    // 장르 내 재생 횟수가 높은 수 최대 2개 수록(단, 하나일경우 하나 수록)
    for(auto &genre : sorted_genres)
    {
        // 장르 내에서 정렬
        auto &songs = genre_plays[genre.first];
        sort(songs.begin(), songs.end(), comparePlay);
        
        // 1개일 경우 1개 출력
        for(int i=0; i<min(2,(int)songs.size()); ++i)
        {
            answer.push_back(songs[i].first);
        }
    }
    
    
    return answer;
}